File: src/main/java/allegro/movieplatform/service/MovieServiceImp.java
package allegro.movieplatform.service;

import allegro.movieplatform.model.Movie;
import allegro.movieplatform.repository.MovieRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import java.util.List;

@Service
public class MovieServiceImp implements MovieService {

    private final static String BASE_URL = "localhost:8080/movies/";

    @Autowired
    private MovieRepository repository;

     public ResponseEntity<List<Movie>> getAllMovies() {
        List<Movie> movies = repository.findAll();
        return new ResponseEntity<>(movies,HttpStatus.OK);
    }

    public ResponseEntity<Void> addMovie(Movie movie) {
        repository.save(movie);
        HttpHeaders responseHeaders = new HttpHeaders();
        String location = BASE_URL + movie.getTitle();
        responseHeaders.set("Location", location);
        return new ResponseEntity<>(responseHeaders, HttpStatus.CREATED);

    }

    public ResponseEntity<Movie> getMovieByTitle(String title){
         Movie movie = repository.findMovieByTitle(title);
         if (movie==null) {
             return new ResponseEntity<>(HttpStatus.NOT_FOUND);
         }
         return new ResponseEntity<>(movie, HttpStatus.OK);
    }
}

File: kotlin/src/main/kotlin/milunas/allegromovies/service/implementation/MovieServiceImp.kt
package milunas.allegromovies.service.implementation

import milunas.allegromovies.model.Movie
import milunas.allegromovies.repository.MovieRepository
import milunas.allegromovies.service.MovieService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
class MovieServiceImp(private val movieRepository: MovieRepository) : MovieService{

    override fun getAllMovies() : List<Movie> {
        return movieRepository.findAll()
    }

    override fun getMovieById(id: Long): Movie {
        return movieRepository.findById(id).get()
    }

    override fun addMovie(movie: Movie) : Movie  {
        if(isMoviePresent(movie))
            throw RuntimeException("Movie " + movie.title + " already in database")

        return movieRepository.save(movie)
    }

    private fun isMoviePresent(movie: Movie) : Boolean{
        return movieRepository.findByTitle(movie.title).isPresent
    }

}